Docs: Add logs data source guide (#24423)
grafana/grafana: docs/sources/developers/plugins/data-sources-for-logs.md

+++
title = "Build a logs data source"
type = "docs"
+++

> # Build a logs data source

Reviewer: Please change the file name to match the title.

This guide explains how to build a logs data source plugin.

Data sources in Grafana supports both metrics and log data. The steps to build a logs data source are largely the same as for a metrics data source. This guide assumes that you're already familiar with how to [Build a data source plugin]({{< relref "../../../../../tutorials/build-a-data-source-plugin.md" >}}) for metrics.

> 1. **Enable logs support**.

Reviewer: Periods should be inside the bolding. In most cases, associated punctuation should be formatted the same as the text.

That said, I think making these these steps AND bold is overkill. What about an H2 to hold both of them and call out the task, then make them both H3? Either that or change both steps to h2.

>    Let Grafana know that your data source plugin can return log data, by adding `"logs": true` to the [plugin.json]({{< relref "metadata.md" >}}) file.

Reviewer: ```suggestion
   Tell Grafana that your data source plugin can return log data by adding `"logs": true` to the [plugin.json]({{< relref "metadata.md" >}}) file.
```

   ```json
   {
     "logs": true
   }
   ```

1. **Construct the log data**.

>    Just like for metrics data, Grafana expects your plugin to return log data as a [data frame]({{< relref "data-frames.md" >}}). Grafana considers any data frame with at least one time field and one text field to be log data.

Reviewer: This is not telling me what to do. I kind of infer that I need to put the log data in a data frame? If so, you should state that.

   ```ts

>    return new MutableDataFrame({

Reviewer 2: MutableDataFrame is only useful if you actually mutate :)  so maybe an example like this is better?

```

    const frame = new MutableDataFrame({
      refId: 'a',
      fields: [
        { name: 'time', type: FieldType.time },
        { name: 'line', type: FieldType.string },
      ],
    });
    frame.add({ time: 12345, line: 'user registered' });
    frame.add({ time: 12345, line: 'user logged in' });
```

     refId: query.refId,
     fields: [
       {
         name: 'time',
         type: FieldType.time,

>          values: [from, to]

Reviewer 3: The from/to is kinda confusing, looks like it should be a range, but it's just timestamp1, timestamp2 basically

Author: This is from the grafana/toolkit scaffold. I'll change to hard-coded timestamp 👍

       },
       {
         name: 'content',
         type: FieldType.string,
         values: ['user registered', 'user logged in']
       },
     ],
   });
   ```

> That's all you need to start returning log data from your data source. Go ahead and try it out in [Explore]({{< relref "../../features/explore.md#explore" >}}), or by adding a [Logs panel]({{< relref "../../features/panels/logs.md#logs-panel" >}}).

Reviewer: ```suggestion
That's all you need to start returning log data from your data source. Go ahead and try it out in [Explore]({{< relref "../../features/explore.md#explore" >}}) or by adding a [Logs panel]({{< relref "../../features/panels/logs.md#logs-panel" >}}).
```

Congratulations, you just wrote your first logs data source plugin! Next, let's look at a couple of features that can improve the experience for the user.

> 

Reviewer: ```suggestion
```

> ## Labels

Reviewer: I suggest changing this heading and the next to "Verb nouns" or "Verb a noun" to maintain consistency with the earlier headings/steps.

To help filter log lines, many log systems let you query logs based on metadata, or _labels_.

You can add labels to a stream of logs by setting the [labels](../../packages_api/data/field.md#labels-property) property on the [Field]({{< relref "../../packages_api/data/field.md" >}}).

> **Example**:

Reviewer: See earlier comments about punctuation.

```ts
return new MutableDataFrame({
  refId: query.refId,
  fields: [
    {
      name: 'time',
      type: FieldType.time,
      values: [from, to]
    },
    {
      name: 'content',
      type: FieldType.string,
      values: ['user registered', 'user logged in'],
      labels: { filename: "file.txt" }
    },
  ],
});
```

## Parsed fields

You can add additional information about each log line by adding more data frame fields.

> If a data frame has more than one text field, Grafana assumes the first field in the data frame to be the actual log line. Any subsequent text fields are treated as [parsed fields]({{< relref "../../features/explore.md#labels-and-parsed-fields" >}}).

Reviewer 3: tbh I don't even know right now :D. We do that for elastic but not sure if also for other DS but if you checked then it makes sense.

Author: That's the behavior I'm seeing at least :P

> If a data frame has more than one text field, Grafana assumes the first field in the data frame to be the actual log line. Any subsequent text fields are treated as [parsed fields]({{< relref "../../features/explore.md#labels-and-parsed-fields" >}}).

Reviewer: ```suggestion
If a data frame has more than one text field, then Grafana assumes the first field in the data frame to be the actual log line. Any subsequent text fields are treated as [parsed fields]({{< relref "../../features/explore.md#labels-and-parsed-fields" >}}).
```

Reviewer: If you use an "if," also include a "then." This helps people with lower-level English skills.
